### ISSUE TITLE
Custom hook gives linting warning "eslint-plugin-valtio": "^0.4.4"

### ISSUE THREAD

**Issue by ndrean** [pmndrs/eslint-plugin-valtio#28]
Follow-up of [this discussion](https://github.com/pmndrs/valtio/discussions/161)

Linter complains when you mutate a snap in a custom hook (not within a React component) to use it in a component.

```js
const state = proxy({ count: 1 })

const useDoubled = () => {
  const snap = useSnapshot(state)
  return snap.count * 2
                ^^^^ Better to just use proxy state.eslint(valtio/state-snapshot-rule)
}

const Component = ()=> {
   const doubled = useDoubled()
  return <>{doubled}
}
```

Other example where linter complains and it shouldn't:

```js
const Component = () => {
  const snap = useSnapshot(store)
 const double = snap*2
                 ^^^^ Better to just use proxy state.eslint(valtio/state-snapshot-rule)
 return <>{double}</>
} 
```

**Comment by dai-shi** [comment]
Yeah, I know we never implemented this.

Can you check if a function has `use` prefix for hooks or returns JSX for components? I don't think we need to track the expressions.

**Comment by barelyhuman** [comment]
So, we already check hooks and assignments 
example. 

this is detected right now
```js
const snap = useSnapshot()
```

also , this was added in one of the PR's where we wanted to let the user to not re-declare the snap 
```js
const {b} = useSnapshot()
const a = ()=>{
b =  b * 1
}
```

so the current conflict is because of the above implementation of write checks, I can modify them for fixing this particular reported case but then it won't work for all cases where a new variable is being use to derive state. 

Hence, the need for thinking it out, I'm writing the test cases right now that'll currently fail with this conflict

**Comment by dai-shi** [comment]
I'm not sure how `const {doubled} = {doubled: snap*2}` can cause the issue, but I don't understand how it uses the syntax tree. There seems to be a simpler way, but I might be missing something. So, I'll wait.

**Comment by barelyhuman** [comment]
here's the syntax tree for just a simple variable declaration
<img width="1440" alt="Screenshot 2022-07-20 at 2 08 12 PM" src="https://user-images.githubusercontent.com/43572006/179937389-f2ea054a-2f07-4efe-bcab-f8115b4f0614.png">

Here's one for a destructured one. 
<img width="1440" alt="Screenshot 2022-07-20 at 2 08 07 PM" src="https://user-images.githubusercontent.com/43572006/179937444-39ac9e1f-3cdb-4c91-aeb0-f00c19f2a686.png">

If we have to find out if the variable declared is actually a snapshot or a derived state from snapshot, that would need a lot more computation than I'd like to have in a eslint plugin. 

> There seems to be a simpler way

Thinking about it, won't be making drastic changes to the current design

**Comment by ndrean** [comment]
I add another use case with the `useEffect` dependencies where the linting is confusing

```js 
const idxStore = proxy({index: 1})

1 const Fetch = () => {
2   const { index } = useSnapshot(idxStore);

3   const [users, setUsers] = React.useState([]);

4   React.useEffect(() => {
5     const getUsers = async (id) => {
6       fetch(id) blabla...;
7       setUsers(...)
8     };
9     getUsers(index);
               ^^^Better to just use proxy state
10  }, [index]
       ^^^Better to just use proxy state
11 );
12  return (...
```

I believe that using the snap `index` is ok since `useEffect` runs at the end, isn't it?

Now I believe the combo `useState`+ `useEffect` can be advantageously replaced by a derivation **+ suspense**.

```js
const derUsers = derive({
   liveUsers= async (get => await  fetch(get(idxStore).index) blabla...)
})

const Fetch  = () => {
  const { liveUsers } = useSnapshot(derUsers)
  return <>{liveUsers}</>
}

<Suspense><Fetch /></Suspense>
```

**Comment by barelyhuman** [comment]
@ndrean  replace `React.useEffect` with just `useEffect` and that should work with the linter, the linter doesn't handle dot expressions right now, but yeah, i'll have to add support for that in both cases so that should fix this for you as well

**Comment by ndrean** [comment]
No change!

**Comment by barelyhuman** [comment]
Okay... 
we have test cases for that , that would've broken for that if it wasn't working. 
I'll check either way, 
Thanks for reporting

**Comment by ndrean** [comment]
But what is the answer? Is it correct to use the snap as a dependency in the end?

**Comment by barelyhuman** [comment]
Yeah, you can use snap if it's defined in the `useEffect` and `useCallback` deps. 

You still shouldn't write to it but reading shouldn't fire it.
